src/player.cpp

```diff
@@ -4678,15 +4678,22 @@ bool player::is_dead_state() const
 
 void player::on_dodge( Creature *source, int difficulty )
 {
-    if( difficulty == INT_MIN && source != nullptr ) {
-        difficulty = source->get_melee();
-    }
+    // Each avoided hit consumes an available dodge
+    // When no more available we are likely to fail player::dodge_roll
+    dodges_left--;
 
-    if( difficulty > 0 ) {
-        practice( skill_dodge, difficulty );
-    }
+    // Even if we are not to train still call practice to prevent skill rust
+    practice( skill_dodge, std::max( difficulty, 0 ) );
 
     ma_ondodge_effects();
+
+    // For adjacent attackers check for techniques usable upon successful dodge
+    if( source && square_dist( pos(), source->pos() ) == 1 ) {
+        matec_id tec = pick_technique( *source, false, true, false );
+        if( tec != matec_id( "tec_none" ) ) {
```

Comment:
This `tec_none` could be wrapped up like it is at the top of melee.cpp:
```
static const matec_id tec_none( "tec_none" );
```

That would be more "proper" than recreating it on every dodge.